### PR TITLE
[WIP] XYX qubit decomp

### PR DESCRIPTION
**Pull request by ecboghiu**
**Context:** Solving issue https://github.com/PennyLaneAI/pennylane/issues/3557

**Description of the Change:**

Following the instructions in https://github.com/PennyLaneAI/pennylane/issues/3557, added a function `xyx_decomposition` in `single_qubit_unitary.py` which implements the decomposition. Also added tests which pass in `test_decompositions.py` using as template the tests for the ZYZ decomposition. The tests run on a random dense unitary, a few Pauli matrices and importantly, also de Identity. The identity triggers the if conditional in the implementation. 

*Details of implementation*
Writing out explicitly the decomposition:
```math
U=e^{i\gamma} e^{-i\phi \sigma_x /2}  e^{-i\theta \sigma_y /2} e^{-i\lambda \sigma_x /2} = e^{i\gamma} \begin{bmatrix}
 \cos \left(\frac{\theta }{2}\right) \cos \left(\frac{\lambda +\phi }{2}\right)+i \sin \left(\frac{\theta }{2}\right) \sin \left(\frac{\lambda -\phi }{2}\right) & -\sin \left(\frac{\theta }{2}\right) \cos \left(\frac{\lambda -\phi }{2}\right)-i \cos \left(\frac{\theta }{2}\right) \sin \left(\frac{\lambda +\phi }{2}\right) \\
 \sin \left(\frac{\theta }{2}\right) \cos \left(\frac{\lambda -\phi }{2}\right)-i \cos \left(\frac{\theta }{2}\right) \sin \left(\frac{\lambda +\phi }{2}\right) & \cos \left(\frac{\theta }{2}\right) \cos \left(\frac{\lambda +\phi }{2}\right)-i \sin \left(\frac{\theta }{2}\right) \sin \left(\frac{\lambda -\phi }{2}\right)
\end{bmatrix}
```
We choose $\gamma$ such that $e^{-i\gamma}U=U_{\mathrm{det1}}$ has the form of the matrix above, which has determinant equal to one. Then we find $\phi$ and $\lambda$ from:

$\frac{\lambda + \phi}{2}= \mathrm{atan2}(-\mathrm{Im}(U_{\mathrm{det1}}(1, 2)), \mathrm{Re}(U_{\mathrm{det1}}(1, 1))$
$\frac{\lambda - \phi}{2}= \mathrm{atan2}(\mathrm{Im}(U_{\mathrm{det1}}(1, 1)), -\mathrm{Re}(U_{\mathrm{det1}}(1, 2))$

For $\theta$ we need to be careful to avoid 0/0:

If $\mathrm{sin}\frac{\lambda + \phi}{2} \neq 0$:
$\theta=2 \mathrm{acos}(-\mathrm{Im}(U_{\mathrm{det1}}(1, 2))/\mathrm{sin}\frac{\lambda + \phi}{2})$ 
else:
$\theta=2 \mathrm{acos}(\mathrm{Re}(U_{\mathrm{det1}}(1, 1))/\mathrm{cos}\frac{\lambda + \phi}{2})$.

**Feedback requested:** I wasn't sure how to return the Operations. The format I chose is as a list of 4 gates. The first one is the identity times the global phase, and the other 3 are the XYX. For example:
```
>>> U = np.array([[-0.28829348-0.78829734j,  0.30364367+0.45085995j],
                  [ 0.53396245-0.10177564j,  0.76279558-0.35024096j]])
>>> decomp = xyx_decomposition(U, 0)
>>> decomp
[(0.38469215914523336-0.9230449299422961j)*(Identity(wires=[0])),
        RX(0.45246583660683803, wires=[0]),
        RY(1.3974974118006183, wires=[0]),
        RX(-1.7210192479534632, wires=[0])]
```
Is this appropriate?

**Benefits:** More decomposition options.

**Possible Drawbacks:** This solution uses an analytical formula which has an edge case of 0/0. This is detected through an if conditional. I'm not sure if this can have a drawback or not. An alternative might be using the [Qiksit XYX implementation](https://qiskit.org/documentation/_modules/qiskit/quantum_info/synthesis/one_qubit_decompose.html#OneQubitEulerDecomposer) which relies on $RX(\phi)RY(\theta)RX(\lambda)=H RZ(\phi)RY(-\theta)RZ(\lambda) H$ to compute the rotation angles. 

**Related GitHub Issues:** https://github.com/PennyLaneAI/pennylane/issues/3557

### CODE REVIEW

**Comment by ecboghiu** [comment]
I tried to modify the changelog but then I got a conflict. I reverted the commit and tried to update but it seems like I cannot. I'll wait for someone to review before modifying the changelog. This would be the addition:

```
* Implemented the XYX single-qubit unitary decomposition. 
  [(#3628)](https://github.com/PennyLaneAI/pennylane/pull/3628) 
```

File: pennylane/transforms/decompositions/single_qubit_unitary.py
Comment (rmoyard) [inline pennylane/transforms/decompositions/single_qubit_unitary.py:123]:
```suggestion
        U (array[complex]): A 2 x 2 unitary matrix.
```

File: pennylane/transforms/decompositions/single_qubit_unitary.py
Comment (rmoyard) [inline pennylane/transforms/decompositions/single_qubit_unitary.py:166]: That's a great trick to add the globaphase! Could make it optional though? I think most hardware would not understand it. You could introduce a `global_phase` boolean arg and turn it off as default.

File: pennylane/transforms/decompositions/single_qubit_unitary.py
Comment (rmoyard) [inline pennylane/transforms/decompositions/single_qubit_unitary.py:148]: You have a utility private function in the file that you can use to convert to su2.

**Review by rmoyard — COMMENTED** [review]
Thanks @ecboghiu it looks that you are on a good way! I've left a couple of suggestions, let me know if you have more questions.

For the changelog, you should pull this pr after clicking the update button and you should be able to not have any conflict. Let me know if the issue persists.

File: pennylane/transforms/decompositions/single_qubit_unitary.py
Comment (ecboghiu) [reply pennylane/transforms/decompositions/single_qubit_unitary.py:148]: I didn't use it at first because that function didn't return the phase, and I didn't want to modify it. But I think its a better idea to use it and add to it the flag `return_global_phase`, unless you think otherwise.

File: pennylane/transforms/decompositions/single_qubit_unitary.py
Comment (ecboghiu) [reply pennylane/transforms/decompositions/single_qubit_unitary.py:166]: I've added the flag `return_global_phase` (and also modified `_convert_to_su2` to accept it). Changed the order in the list s.t. the global phase operators is the last, if returned.

**Comment by ecboghiu** [comment]
I implemented your suggestions and adapted the tests. I also managed to update the changelog with no conflicts.

File: doc/releases/changelog-dev.md
Comment (rmoyard) [inline doc/releases/changelog-dev.md:170]: Don't forget to add yourself as contributor at the bottom of the file :)

**Review by rmoyard — COMMENTED** [review]
Thank you for the changes, it looks great! Two remaining things before approval:

1. Add yourself to the changelog
2. Add `xyx_decomposition` to the `_init_` transforms,  to make it top level and also add it to the doc. Look at the zyz case if you need.

**Comment by ecboghiu** [comment]
> 2. ... l and also add it to the doc. Look at the zyz case if you need.

In the [documentation guide](https://docs.pennylane.ai/en/stable/development/guide/documentation.html) they mention not to manually modify the module listing. I tried running `sphinx-apidoc -f -o ../pennylane/ .` in the `./doc` folder but I didn't manage to get `pennylane/transforms/__init__.py` to update the docstring to include `~transforms.xyx_decomposition` where appropriate. So I did it manually. Now when compiling the HTML documentation with `make html` the function appears correctly in the documentation. Maybe there was a smarter way to update the documentation, but what I did seems to work.

File: pennylane/conf.rst
Comment (rmoyard) [inline pennylane/conf.rst:1]: You don't need to add this file.

File: pennylane/modules.rst
Comment (rmoyard) [inline pennylane/modules.rst:1]: Same for this one.

File: pennylane/transforms/__init__.py
Comment (rmoyard) [inline pennylane/transforms/__init__.py:83]: This is sufficient for building the documentation of the function.

**Review by rmoyard — COMMENTED** [review]
Thanks @ecboghiu ,  it looks great! You don't need to add the two rst files to show the documentation. Once it is built you can simply click on details: 
here and it will show you the build. Happy to approve after you remove the rst files :+1:

**Review by rmoyard — APPROVED** [review]
Thank you @ecboghiu, great job :100: Very nice contribution!

**Pull request merged**
